README
PHP framework

CHANGELOG.txt
TODO:
	Add method of selecting module load order
	Build a small website to try everything out.



+  : ADDED
*  : CHANGED
-  : REMOVED
!  : INFO

--

BETA 4 !INPROGRESS! (2012-05-26)
	+ Added functions (generate_download_url, get_path)
	+ Added ability to use external javascripts
	+ Added ability to use inline javascript
	+ Added ability to catch url as arguments instead of comming to a 404 if template file does not exist
	+ Modules now has it's own folders to work with for file includes and other things.
	+ Added module (tiny_mce)
	+ Added basic html/css design

	* Small changes to javascript include handle because of external/inline javascript
	* Template file of current page is now available as $content in main.tpl.php. Just print it.
	* Modules now have access to the $fw object for database actions
	* All functions in "core/functions" now uses underscore instead of camelcase
	* The "$vars['style']" and "$vars['script']" arrays now gets cleaned after "page_preprocess_all"
	* Javascript includes now has to be written with the extension in the entire framework
	* Change module folder name from "application/classes" to "application/modules"
	* Change module filename from ex. "tiny_mce.class.php" to tiny_mce.php
	
	- Removed module (replacer)


---


BETA 3 (2012-05-05)
	+ Added automatic class loading
	+ Added example class file
	+ Added function (timeago)

	* Optimized $vars array handling
	* Slightly changed the preprocesses handling of the $vars array
	* Updated jQuery 1.7.1 -> 1.7.2

	! Average load time without cache including 1 class (local): 0.01 Sec


---


BETA 2v2 (2012-02-25)
	+ Added function (arg)
	+ Added security layer in htaccess file. (Only allows certain files to be accessed)

	* Vastly improved page handling in htaccess file (Thanks to GOOSK)
	* Some changes in template file handling because of ^
	* Renamed custom template functions from ex. "page_all" to "page_preprocess_all"
	* New output array in template files. ($page_all['name'] and $page_mypage['dates']) to a combined: ($vars['name'] and $vars['dates'])
	* Renamed template files to .tpl.php instead of .php only


---


BETA 2 (2012-02-02)
	+ Added functions (validateEmail, redirect)
	
	* Updated jQuery 1.6.0 -> 1.7.1
	* Handling of template functions. Now uses references instead of return value
	* Added $randomString1, $randomString2 to variable unset

	
---
	
	
BETA 1 (2011-10-16)
	+ Added basic cache
	
	* Lots of changes everywhere
	
	- Removed uneccecary code

index.php
<?php // 2012-05-25
	session_start();
	define("INDEX", true);
	ob_start();

	include_once('settings.php');
	include_once('core/functions/functions.php');
	$page = end(explode('/',$_GET['args']));
	$args = trim(implode('.',explode('/',$_GET['args'])),'.');
	
	if(empty($args))	$args = 'start';
	$pages = $args;

	if($cache === TRUE) {
		$cachefile = base64_encode($pages.'.html');
	}

	if($cache === TRUE && file_exists('cache/'.$cachefile) && (time() - $cachetime < @filemtime('cache/'.$cachefile))) {
		// LOAD CACHE FILE
		include('cache/'.$cachefile);
	} else {
		
		// PageFunction to run
		if(!isset($pageFunction)) $pageFunction = 'page_preprocess_'.str_replace('.','_',$args);

		// Framework class include
		require_once('core/classes/framework.class.php');

		// Init SQL
		$mysqli	= new Mysqli($framework['dbHost'],$framework['dbUser'],$framework['dbPass'],$framework['dbData']);
		if($mysqli->connect_errno) {
			printf("Mysql Error: %s\n", $mysqli->connect_error);
			die();
		}

		// Init framework
		$fw = new Framework(array('mysqli' => $mysqli, 'hash' => array($randomString1,$randomString2)));

		// Load template functions
		require_once('application/functions.php');

		// Data storage
		$vars = array();
		$content  = '';
		//$catch_args = FALSE;
		$framework['script']['inline'] = array();

		// Get all classes to load
		$classes = $fw->load_classes('application/modules/');

		foreach($classes['file'] AS $k => $class) {
			if(!empty($class)) {
				// Include classes
				require_once($class);
				if(!class_exists($classes['name'][$k])) { 
					throw new ErrorException('Class Not Found !');
				} else { 
					// Create instances
					${$classes['name'][$k]} = new $classes['name'][$k]($fw);
					if(method_exists(${$classes['name'][$k]},'returner')) {
						// Run all returners
						${$classes['name'][$k]}->returner($vars);
					}
				}
			}
		}

		// Load page_preprocess_all function if it exists
		if(function_exists('page_preprocess_all')) page_preprocess_all($vars, $fw);

		// Pick up styles and script if set in template function
		if(isset($vars['style'])) $framework['style']	= array_merge($framework['style'],$vars['style']);
		if(isset($vars['script'])) $framework['script']	= array_merge($framework['script'],$vars['script']);
		
		$vars['style'] = array();
		$vars['script'] = array();
		
		// Catch arguments
		if(isset($vars['catch_args'])) {
			foreach($vars['catch_args'] AS $arg) {
				$file = str_replace('_', '.', $arg);
				if($file != $pages) {
					if(strpos($pages, $file) !== FALSE) {
						if(file_exists('application/pages/'. $file .'.page.tpl.php')) {
							$pageFunction = 'page_preprocess_'.$arg;
							$vars['catch_args'] = explode('.', str_replace($file.'.','',$pages));
							$pages = $file;
						}
					}
				} else {
					$vars['catch_args'] = array();
				}
			}
		}
		
		// Load page function if it exists
		if(function_exists($pageFunction)) $pageFunction($vars, $fw);

		// Run all modifiers
		for($k=0;$k<sizeof($classes['name']);$k++) {
			if(method_exists(${$classes['name'][$k]},'modifier')) {
				${$classes['name'][$k]}->modifier($vars);
			}
		}

		// Pick up styles and script if set in template function
		if(isset($vars['style']))	$framework['style']	= array_merge($framework['style'],$vars['style']);
		if(isset($vars['script']['inline']))	$framework['script']['inline']	= array_merge($framework['script']['inline'],$vars['script']['inline']);
		if(isset($vars['script']['inline']))	unset($vars['script']['inline']);
		if(isset($vars['script']))	$framework['script']	= array_merge($framework['script'],$vars['script']);

		// Framework variables
		$pageName	= $framework['pageName'];
		$title			= $page;

		if(isset($vars['title'])) $title = $vars['title'];

		$styles	= $fw->getStyles($framework['style']);		// Returns styles
		$scripts	= $fw->getScripts($framework['script']);	// Returns scripts

		// Everything except template is now loaded. Unset unnecessary and sensitive variables
		unset($framework,$key,$vars['script'],$vars['style'],$vars['catch_args'],$pageFunction,$GLOBALS,$dta,$randomString1,$randomString2);

		// Start loading templates
		ob_start();
		if(file_exists('application/pages/'. $pages .'.page.tpl.php')) {
			include('application/pages/'. $pages .'.page.tpl.php');
		} else {
			include('application/pages/notfound.tpl.php');
		}
		$content = ob_get_clean();
		
		require_once('application/main.tpl.php');
		$mysqli->close();

		if($cache === TRUE) {
			echo "<!-- Cache updated ".date('jS F Y H:i')." -->";
			$fp = fopen('cache/'.$cachefile, 'w');
			fwrite($fp, ob_get_contents());
			fclose($fp);
			ob_end_flush();
		}
	}

?>